src/selectors/columnData.js

```diff
@@ -0,0 +1,148 @@
+    forEach(children, (columnGroupElement, index) => {
+      elementTemplates.groupHeader.push(columnGroupElement.props.header);
+
+      React.Children.forEach(columnGroupElement.props.children, (child) => {
+        const column = extractPropsFromElement(child);
+        column.groupIdx = index;
+        columnProps.push(column);
+        extractTemplatesFromElement(elementTemplates, child);
+      });
+    });
+
+    return {
+      columnGroupProps,
+      columnProps,
+      elementTemplates,
+      useGroupHeader: true,
+    };
+  }
+
+  // Use a default column group
+  forEach(children, (child) => {
+    columnProps.push(extractPropsFromElement(child));
+    extractTemplatesFromElement(elementTemplates, child);
+  });
+  return {
+    columnGroupProps: [],
+    columnProps,
+    elementTemplates,
+    useGroupHeader: false,
+  };
+}
+
+function fixColumnPropsWithColumnGroupProps(columnProps, columnGroupProps) {
+  const columnGroups = [];
+  let columnsWithDefinedGroupIndex = 0;
+
+  forEach(columnProps, ({ groupIdx }) => {
+    if (groupIdx !== undefined && groupIdx !== null) {
+      columnGroups[groupIdx] = true;
+      columnsWithDefinedGroupIndex++;
+    }
+  });
+
+  invariant(columnGroups.length === (columnGroupProps || []).length,
+    'column group data must fully and only contain every column groups');
+
+  invariant(columnsWithDefinedGroupIndex === columnProps.length,
+    'group index if specified must be given for every column');
+
+  forEach(columnProps, (columnProp) => {
+    columnProp.fixed = columnGroupProps[columnProp.groupIdx].fixed;
+    columnProp.fixedRight = columnGroupProps[columnProp.groupIdx].fixedRight;
+  });
+}
+
+/**
+ * Uses columnData to extract column props and cell rendering templates
+ *
+ * @param {!Array.<columnDefinition>} columnData
+ * @param {!Array.<columnDefinition>} columnGroupData
+ * @return {{
+ *   columnGroupProps: !Array.<columnDefinition>,
+ *   columnProps: !Array.<columnDefinition>,
+ *   elementTemplates: !Object.<string, Array>,
+ *   useGroupHeader: boolean,
+ * }}
+ */
+function getColumnPropsFromData(columnData, columnGroupData) {
+  const columnProps = extractPropsFromData(columnData);
+  const columnGroupProps = extractPropsFromData(columnGroupData);
+
+  const elementTemplates = extractTemplatesFromData(columnData, columnGroupData);
+
+  fixColumnPropsWithColumnGroupProps(columnProps, columnGroupProps);
+
+  return {
+    columnGroupProps,
+    columnProps,
+    elementTemplates,
+    useGroupHeader: columnGroupProps.length > 0,
+  };
+}
+
+function getColumnData(allowColumnVirtualization, columnData, columnGroupData, childComponents) {
+  // use columnData to directly get the column data
+  if (allowColumnVirtualization) {
+    return getColumnPropsFromData(columnData, columnGroupData);
+  }
+
+  // use legacy API (React Children) to get the column data
+  return convertColumnElementsToData(childComponents);
+}
+
+export default shallowEqualSelector([
```

Comment:
Made this into a selector.

The aim is that we'll only extract columns if there's a change. 
Unfortunately for the old API, `React children` is a new instance at each render, and hence reselect won't work. But this does work for our data driven approach.

This means that controlled scrolling will be much faster for the new API,

Reply:
Nice.  Good idea to convert to data first and then do all the needed changes off that.